### PR TITLE
Adding a docker file to help test for missing package.xml dependencies

### PR DESCRIPTION
**Pull request by crdelsey**
## Description of contribution in a few bullet points

* Added a docker file that builds ROS2, the navstack dependencies and navigation 2 in one big workspace without sourcing any setup.bash files. This catches any missing ROS2 dependencies in our package.xml files.
* Also added a Pre-release checklist doc.

---

## Future work that may be required in bullet points

 * There is a lot of duplication among our docker files that needs to be fixed. I don't know how to do that yet.
* This docker file starts from `crystal-ros-core` to get the correct apt sources file and the basic dependencies that ROS2 needs, however `crystal-ros-core` installs a bunch of ros2 packages we don't need since we are building everything from source anyhow.
* The `crystal-ros-core` base image probably needs to be an argument, so we can change it based on which branch we are targeting.

### CODE REVIEW

**Comment by ruffsl** [comment]
> Added a docker file that builds ROS2, the navstack dependencies and navigation 2 in one big workspace without sourcing any setup.bash files. This catches any missing ROS2 dependencies in our package.xml files.

I'm not sure I understand. This seem to be using rosdep to install all dependencies declared in workspace packages, how does using one workspace instead of a layered workspace for upstream packaged dependencies change things? I think we would need to call rosdep for only the navigation packages, and the upstream ros2 packages that and dependent, not necessary for all packages in the ros2 master .repo.

> This docker file starts from crystal-ros-core to get the correct apt sources file and the basic dependencies that ROS2 needs, however crystal-ros-core installs a bunch of ros2 packages we don't need since we are building everything from source anyhow.

Is this big issue? I could probably add a new tag like `osrf/ros2:devel` that only configures that sources.list and common build tools. But I think ros2 is missing a ros install generator like ros1 had that could help resolve your issue at hand. Also, perhaps this would build faster if we limited the build by using `--pacakges-up-to <list of navigation2 packages>`


> There is a lot of duplication among our docker files that needs to be fixed. I don't know how to do that yet.

I think introducing another Dockerfile would exacerbates this. I've recommend templating our Dockerfile pipeline using docker build arguments to generalize against any release, specially for the use case in our CI and target release branch PRs. I'm little blocked however in doing this by upstream issues however:

https://discuss.circleci.com/t/conditioning-docker-image-via-base-branch-name-from-pr/28426/6

**Comment by crdelsey** [comment]
> This seem to be using rosdep to install all dependencies declared in workspace packages, how does using one workspace instead of a layered workspace for upstream packaged dependencies change things?

Let me make a simple example to explain what I believe happens. Let's say ROS2 consists of 2 packages: `alpha` and `beta` and Nav2 consists of the package `gamma`.

Our typical build structure looks like this:
image

We normally build `ros2_ws` and then source `ros2_ws/install/setup.bash` before building the `nav2_ws`. When we source that file, we get every package in `ros2_ws` added to the environment.  eg. `CMAKE_PREFIX_PATH=/nav2_build/ros2_ws/install/alpha:/nav2_build/ros2_ws/install/beta`
All other build related environment variables are also set in a similar way.

Let's say `gamma` depends on `alpha` and `beta` but we forgot to include `beta` in `gamma`'s package.xml. `gamma` frequently still builds correctly because CMake finds all relevant dependencies in the search path.

In this new docker file, it's the equivalent of building at the `nav2_build` level.
image

In this case, we never source a `setup.bash` before running colcon, so when it tries to build `gamma`, the build fails because CMake can't find the `beta` dependencies. 

-----

I believe `colcon` creates a temporary environment for each package it builds by taking the current environment, and then adding on the paths to any dependencies listed in package.xml. In the second case above, we fail because we start with a clean environment and colcon only adds in the `alpha` dependency.

In the first case, we succeed because `alpha` and `beta` are already in the environment, so it doesn't matter what colcon adds or doesn't add.

---- 

>  This seem to be using rosdep to install all dependencies declared in workspace packages

There are two classes of dependencies here: ROS2 dependencies and third party dependencies. The one workspace build presented here is only meant to catch missing ROS2 dependencies. We frequently miss the ROS2 dependencies in the package.xml files because they were already installed as part of the `ros2_ws` build. Then we fail when OSRF does their nightly build because they are building each package with an independent environment.

---- 
> Is this big issue? I could probably add a new tag like osrf/ros2:devel

It's not a big problem. We are just pulling more into the docker image than really is needed. 

My real concern is that some pre-installed packages will leak into the environment causing the build to succeed when it should really fail. I don't think this is the case right now, but could happen because I'm not using the docker the way it is really intended

> Also, perhaps this would build faster if we limited the build by using --pacakges-up-to <list of navigation2 packages>

Yes. I believe you're right. Making that change would require human time, whereas just building everything only takes computer time, so I didn't bother. However, if we incorporate this extra type of build into CI, we should do that.

I'm not sure if this type of build should go in CI, of if it should just stay a manual verification step done prior to release.

Does this explain what I'm trying to do here?

**Comment by ruffsl** [comment]
> Does this explain what I'm trying to do here?

Thanks for the thorough explanation! I think understand now: the checking for missing ros2 depnacies.

> I'm not sure if this type of build should go in CI, of if it should just stay a manual verification step done prior to release.

I think we could either have it a regular cron job for master/release branches or we could have it be a final check in PRs, e.g. before merging, user merging PR first manual approves the branch of the workflow for the PR that tests this, so that not every commit triggers this length build.

https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval

File: doc/process/PreReleaseChecklist.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+### Ensure all dependencies are listed in the package.xml files
+
+We want to ensure that every package has a complete list of its dependencies
+in the `package.xml` file. This can be done by not sourcing any ros `setup.bash` files. Instead we need to build everything as one big repo.
+
+There is a docker file to do that, so run
+
+```bash
+sudo docker build -t nav2:crystal_full  --build-arg http_proxy=http://myproxy.example.com:80  --build-arg https_proxy=http://myproxy.example.com:80 --build-arg ROS2_BRANCH=crystal -f Dockerfile.full_ros_build ./
```
Comment (mkhansen-intel) [inline doc/process/PreReleaseChecklist.md]: Did you get this command and the other one swapped? This one says ROS2_BRANCH=crystal, which I think should be in the next example?

File: doc/process/PreReleaseChecklist.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+```bash
+sudo docker run nav2:crystal colcon test
+```
+
+#### Dashing and newer
+
+For newer releases, run
+
+```bash
+sudo docker run nav2:crystal src/navigation2/tools/run_test_suite.bash
```
Comment (mkhansen-intel) [inline doc/process/PreReleaseChecklist.md]: should this be nav2:master?

File: doc/process/PreReleaseChecklist.md (hunk 1 above)
Comment (crdelsey) [reply doc/process/PreReleaseChecklist.md]: Both dockerfiles need a ROS2_BRANCH argument now. It represents the upstream branch we are targeting or verifying against.

File: doc/process/PreReleaseChecklist.md (hunk 2 above)
Comment (crdelsey) [reply doc/process/PreReleaseChecklist.md]: It's just a tag for the docker build. We could leave it out entirely. I'll change the tags to be branch agnostic. I'll make them nav2:full_build and nav2:rosdep_build

**Pull request merged**